grails-app/controllers/au/org/ala/biocollect/BioActivityController.groovy

```diff
@@ -1801,6 +1804,171 @@ class BioActivityController {
+                if (!result.error) {
+                    render text: result as JSON, contentType: ContentType.APPLICATION_JSON
+                } else {
+                    render text: getMessage(result), status: HttpStatus.SC_INTERNAL_SERVER_ERROR, contentType: ContentType.APPLICATION_JSON
+                }
+            }
+        } else {
+            render text: "Access denied: User not authorised.", status: HttpStatus.SC_UNAUTHORIZED
+        }
+    }
+
+    @Operation(
+            method = "GET",
+            tags = "biocollect",
+            summary = "Generate darwin core archive",
+            description = "Returns a ZIP file containing darwin core archive",
+            parameters = [
+                    @Parameter(
+                            name = "projectId",
+                            in = ParameterIn.PATH,
+                            required = true,
+                            description = "Project id"
+                    )
+            ],
+            responses = [
+                    @ApiResponse(
+                            responseCode = "200",
+                            description = "Successful response with ZIP file",
+                            content = [@Content(mediaType = "application/zip", schema = @Schema(type = "string", format = "binary"))]
+                    ),
+                    @ApiResponse(responseCode = "400", description = "Bad request"),
+                    @ApiResponse(responseCode = "401", description = "Unauthorized"),
+                    @ApiResponse(responseCode = "500", description = "Internal server error")
+            ],
+            security = @SecurityRequirement(name = "auth")
+    )
+    @Path("ws/bioactivity/data/archive/{projectId}")
+    @Produces("application/zip")
+    def getDarwinCoreArchiveForProject(String projectId){
+        log.debug("projectId = ${projectId}")
+
+        String userId = userService.getCurrentUserId()
+        Map project = projectService.get(projectId)
+        if (!userId) {
+            render text: [message: "Access denied: User not authorised."] as JSON, status: HttpStatus.SC_UNAUTHORIZED, contentType: ContentType.APPLICATION_JSON
+        } else if (!projectId) {
+            render text: "Project Id is required.", status: SC_BAD_REQUEST
+        }else if (project.error) {
+            render text: [message: "An error occurred when accessing project."] as JSON, status: HttpStatus.SC_INTERNAL_SERVER_ERROR, contentType: ContentType.APPLICATION_JSON
+        } else {
+            if (projectService.canUserEditProject(userId, projectId, false)) {
+                try {
+                    response.contentType = 'application/zip'
+                    response.setHeader("Content-disposition", "attachment; filename=darwin-core.zip")
+                    activityService.getDarwinCoreArchiveForProject(projectId, response)
+                    response.outputStream.flush()
```

Comment:
add `return null`. There used to be a bug in grails if not added. Not sure if it has been corrected.

Reply:
I've added return null